### PR TITLE
IBP-3899 Refactor Sorted Page Request

### PR DESCRIPTION
**Pull request by abatac**

### CODE REVIEW

File: src/main/java/org/ibp/api/java/impl/middleware/dataset/DatasetServiceImpl.java [hunk 1]
```diff
@@ -311,7 +320,22 @@ void validateStudyDatasetAndInstances(
...
 	@Override
 	public List<ObservationUnitRow> getObservationUnitRows(
-		final int studyId, final int datasetId, final ObservationUnitsSearchDTO searchDTO) {
+		final int studyId, final int datasetId, final ObservationUnitsSearchDTO searchDTO, final Pageable pageable) {
+
+		Pageable convertedPageable = null;
+		if (pageable != null && pageable.getSort() != null) {
+			final Iterator<Sort.Order> iterator = pageable.getSort().iterator();
+			if (iterator.hasNext()) {
+				// Convert the sort property name from termid to actual term name.
```
Comment (darla-leafnode) [inline src/main/java/org/ibp/api/java/impl/middleware/dataset/DatasetServiceImpl.java:329]: 1. Why did you move this logic to lookup term name by term id from Middleware to BMSAPI?
2. Another note is maybe we can find a way to optimize by getting all variables IDS one-off

File: src/main/java/org/ibp/api/java/impl/middleware/study/StudyServiceImpl.java [hunk 2]
```diff
@@ -389,8 +390,8 @@ public long countStudies(final StudySearchFilter studySearchFilter) {
 	}
 
 	@Override
-	public List<StudyDto> getStudies(final StudySearchFilter studySearchFilter) {
-		return this.middlewareStudyService.getStudies(studySearchFilter);
+	public List<StudyDto> getStudies(final StudySearchFilter studySearchFilter, final PageRequest pageRequest) {
```
Comment (darla-leafnode) [inline src/main/java/org/ibp/api/java/impl/middleware/study/StudyServiceImpl.java:393]: Use `final Pageable pageable`

File: src/main/java/org/ibp/api/java/study/GermplasmStudySourceService.java [hunk 3]
```diff
@@ -2,12 +2,14 @@
...
 import org.generationcp.middleware.service.api.study.germplasm.source.GermplasmStudySourceSearchRequest;
+import org.springframework.data.domain.PageRequest;
+import org.springframework.data.domain.Pageable;
 
 import java.util.List;
 
 public interface GermplasmStudySourceService {
 
-	public List<GermplasmStudySourceDto> getGermplasmStudySources(GermplasmStudySourceSearchRequest germplasmStudySourceSearchRequest);
+	public List<GermplasmStudySourceDto> getGermplasmStudySources(GermplasmStudySourceSearchRequest germplasmStudySourceSearchRequest, Pageable pageRequest);
```
Comment (darla-leafnode) [inline src/main/java/org/ibp/api/java/study/GermplasmStudySourceService.java:12]: Use `pageable` parameter name

File: src/main/java/org/ibp/api/java/study/StudyService.java [hunk 4]
```diff
@@ -67,7 +68,7 @@
 
 	long countStudies(StudySearchFilter studySearchFilter);
 
-	List<StudyDto> getStudies(StudySearchFilter studySearchFilter);
+	List<StudyDto> getStudies(StudySearchFilter studySearchFilter, PageRequest pageRequest);
```
Comment (darla-leafnode) [inline src/main/java/org/ibp/api/java/study/StudyService.java:71]: Use `final Pageable pageable`

File: src/main/java/org/ibp/api/rest/dataset/DatasetResource.java [hunk 5]
```diff
@@ -184,26 +188,29 @@
...
+			value = "Number of records per page."),
+		@ApiImplicitParam(name = "sort", allowMultiple = false, dataType = "string", paramType = "query",
+			value = "Sorting criteria in the format: property,asc|desc. ")
+	})
 	public ResponseEntity<ObservationUnitTable> getObservationUnitTable(@PathVariable final String cropname, @PathVariable final String programUUID,
 		@PathVariable final Integer studyId, //
 		@PathVariable final Integer datasetId, //
-		@RequestBody final ObservationUnitsSearchDTO searchDTO) {
+		@RequestBody final ObservationUnitsSearchDTO searchDTO,
+		@ApiIgnore @PageableDefault(page = 0, size = 50) final Pageable pageable) {
```
Comment (darla-leafnode) [inline src/main/java/org/ibp/api/rest/dataset/DatasetResource.java:203]: Why choose 50 as default size?

File: src/main/java/org/ibp/api/rest/labelprinting/ObservationDatasetLabelPrinting.java [hunk 6]
```diff
@@ -281,7 +281,7 @@ public LabelsData getLabelsData(final LabelsGeneratorInput labelsGeneratorInput)
 		studyTransactionsRequest.setTransactionsSearch(transactionsSearch);
 
 		final List<StudyTransactionsDto> studyTransactionsDtos =
-			studyTransactionsService.searchStudyTransactions(labelsGeneratorInput.getStudyId(), studyTransactionsRequest);
+			studyTransactionsService.searchStudyTransactions(labelsGeneratorInput.getStudyId(), studyTransactionsRequest, null);
```
Comment (darla-leafnode) [inline src/main/java/org/ibp/api/rest/labelprinting/ObservationDatasetLabelPrinting.java:284]: How about converting the `Pageable` parameter as `Optional<Pageable>`? That way we wouldn't be passing decreasing null values as parameters if paging is not applicable.

**Review by darla-leafnode — APPROVED** [review]

**Review by darla-leafnode — CHANGES_REQUESTED** [review]
Please check my comments @abatac

File: src/main/java/org/ibp/api/java/impl/middleware/dataset/DatasetServiceImpl.java (hunk 1 above)
Comment (abatac) [reply src/main/java/org/ibp/api/java/impl/middleware/dataset/DatasetServiceImpl.java:329]: I moved that logic from Middleware (fillSearchDto) to ensure that the Pageable's order by value is already converted to the proper term name before passing it to the middleware method. Pageable is an interface so we cannot transform/overwrite that inside the middleware method.

File: src/main/java/org/ibp/api/rest/dataset/DatasetResource.java (hunk 5 above)
Comment (abatac) [reply src/main/java/org/ibp/api/rest/dataset/DatasetResource.java:203]: This should be @PageableDefault(page = 0, size = PagedResult.DEFAULT_PAGE_SIZE).

File: src/main/java/org/ibp/api/java/impl/middleware/study/GermplasmStudySourceServiceImpl.java [hunk 7]
```diff
@@ -6,6 +6,8 @@
 import org.generationcp.middleware.util.CrossExpansionProperties;
 import org.ibp.api.java.impl.middleware.study.validator.StudyValidator;
 import org.ibp.api.java.study.GermplasmStudySourceService;
+import org.springframework.data.domain.PageRequest;
```
Comment (darla-leafnode) [inline src/main/java/org/ibp/api/java/impl/middleware/study/GermplasmStudySourceServiceImpl.java:9]: PageRequest is unused

File: src/main/java/org/ibp/api/java/impl/middleware/inventory/study/StudyTransactionsServiceImpl.java [hunk 8]
```diff
@@ -6,6 +6,8 @@
 import org.ibp.api.java.impl.middleware.study.validator.StudyValidator;
 import org.ibp.api.java.inventory.manager.TransactionService;
 import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.data.domain.PageRequest;
```
Comment (darla-leafnode) [inline src/main/java/org/ibp/api/java/impl/middleware/inventory/study/StudyTransactionsServiceImpl.java:9]: Remove unused import

**Review by darla-leafnode — APPROVED** [review]
